research/fed-ce/jobs/fedce_prostate/app/custom/aggregators/fedce_assembler.py

```diff
@@ -0,0 +1,168 @@
+                    temp += self.fedce_coef[site] * torch.as_tensor(data[site]["model"][name])
+                consensus_grad.append(temp.data.view(-1))
+
+        # flatten for cosine similarity computation
+        consensus_grads_vec = torch.cat(consensus_grad).to("cpu")
+
+        # generate minus gradients and compute cosine similarity
+        for site in site_list:
+            site_grad = []
+            for name in self.fedce_cos_param_list:
+                site_grad.append(torch.as_tensor(data[site]["model"][name]).data.view(-1))
+            site_grads_vec = torch.cat(site_grad).to("cpu")
+            # minus gradient
+            minus_grads_vec = consensus_grads_vec - self.fedce_coef[site] * site_grads_vec
+            # compute cosine similarity
+            fedce_cos_sim_site = (
+                torch.cosine_similarity(site_grads_vec, minus_grads_vec, dim=0).detach().cpu().numpy().item()
+            )
+            # append to record dict
+            self.fedce_cos_sim[current_round][site] = fedce_cos_sim_site
+
+        # compute cos_weights and minus_vals based on the record for each site
+        fedce_cos_weights = []
+        for site in site_list:
+            # cosine similarity
+            cos_accu_avg = np.mean([self.fedce_cos_sim[i][site] for i in range(current_round + 1)])
+            fedce_cos_weights.append(1.0 - cos_accu_avg)
+
+        # normalize
+        fedce_cos_weights /= np.sum(fedce_cos_weights)
+        fedce_cos_weights = np.clip(fedce_cos_weights, a_min=1e-3, a_max=None)
+        fedce_minus_vals /= np.sum(fedce_minus_vals)
+        fedce_minus_vals = np.clip(fedce_minus_vals, a_min=1e-3, a_max=None)
+
+        # two aggregation strategies
+        if self.fedce_mode == "times":
+            new_fedce_coef = [c_w * mv_w for c_w, mv_w in zip(fedce_cos_weights, fedce_minus_vals)]
+        elif self.fedce_mode == "plus":
+            new_fedce_coef = [c_w + mv_w for c_w, mv_w in zip(fedce_cos_weights, fedce_minus_vals)]
+        else:
+            raise NotImplementedError
+
+        # normalize again
+        new_fedce_coef /= np.sum(new_fedce_coef)
+        new_fedce_coef = np.clip(new_fedce_coef, a_min=1e-3, a_max=None)
+
+        # update fedce_coef
+        fedce_coef = {}
+        idx = 0
+        for site in site_list:
+            fedce_coef[site] = new_fedce_coef[idx]
+            idx += 1
+
+        # compute global model update with the new fedce weights
+        global_updates = {}
+        for idx, name in enumerate(global_weights):
+            temp = torch.zeros_like(global_weights[name], dtype=torch.float32)
+            for site in site_list:
+                weight = fedce_coef[site]
+                temp += weight * data[site]["model"][name]
+            global_updates[name] = temp.detach().cpu().numpy()
+
+        meta = {"fedce_coef": fedce_coef}
+
+        dxo = DXO(data_kind=self.expected_data_kind, data=global_updates)
+        for entry in meta.keys():
```

Comment:
You can just use `DOX(..., meta=meta)` here instead of the loop.

Reply:
yes indeed, updated